lib/gears/matcher.lua

```diff
@@ -35,10 +35,45 @@
 local gtable = require("gears.table")
 local gsort = require("gears.sort")
 local gdebug = require("gears.debug")
+local gobject = require("gears.object")
 local protected_call = require("gears.protected_call")
 
 local matcher = {}
 
+--- A rule has been added to a set of matching rules.
```

Comment:
Yes and no.

This isn't what this is used for "in practice". This is used by the `tag` rules to do house keeping on the `name` property of the tag (within the `properties` section of the rule).

It is also used for the `init` tag property, which tells to create this tag on initialization (aka, as soon as the rule is added). When `init` is set to `false`, the tag is only created when the rules match the current state (usually when a new client with a certain class is added).

Finally, the last use case is to build an index of "expected [client] classes" and sort the rules between the ones that "really" need to be looped over and the ones who can be matched directly from an housekeeping directory (this part is in a distant pull request *way* after the much simpler implementation is proposed). This speeds up startup a lot when there is both multiple clients and multiple rules. It has a little bug where it isn't possible to detect when a string is expected to be used with `match`. Tyrannical had it from the start because with such system, having dozens of rules is common, but very few of them really require one-by-one matching.